crates/fs/private/hamt/node.rs

```diff
@@ -169,9 +204,61 @@ where
             }
         }
     }
+
+    #[async_recursion(?Send)]
+    pub(super) async fn remove_value<'a, 'b, B: BlockStore>(
+        self: &'a Rc<Self>,
+        hashnibbles: &'b mut HashNibbles,
+        key: &K,
+        store: &B,
+    ) -> Result<(Rc<Self>, Option<Pair<K, V>>)> {
+        let bit_index = hashnibbles.try_next()?;
+
+        // If the bit is not set yet, return None.
+        if !self.bitmask[bit_index] {
+            return Ok((Rc::clone(self), None));
+        }
+
+        // Otherwise, we go one level deep.
+        let value_index = self.get_value_index(bit_index);
+        Ok(match &self.pointers[value_index] {
+            Pointer::Values(values) => {
+                let mut node = (**self).clone();
+                let value = if values.len() == 1 {
+                    // If there is only one value, we can remove the entire pointer.
+                    node.bitmask.set(bit_index, false);
+                    let pointer = node.pointers.remove(value_index);
```

Comment:
In order to keep determinism, we have to check whether removing this pointer gets us within bucket size range. I.e. when this node contains <= 3 values, we can collapse it into a pointer and instead of linking to it, we merge it into the parent node.